File: scripts/power_twin.py
import csv
import json
import argparse
from dataclasses import dataclass
from typing import List, Dict, Any, Optional
from datetime import datetime, timezone
import os

# --- TC-S constants ---
SOLAR_KWH = 4913.0          # 1 Solar = 4,913 kWh
RAYS_PER_SOLAR = 10000.0    # 1 Solar = 10,000 Rays


@dataclass
class PowerSample:
    time_s: float   # seconds
    power_w: float  # watts


def load_power_trace_csv(path: str) -> List[PowerSample]:
    """
    Load a CSV with at least two columns: time_s, power_w.

    Example:
        time_s,power_w
        0.0, 80
        0.01, 95
        0.02, 90
    """
    samples: List[PowerSample] = []

    with open(path, "r", newline="") as f:
        reader = csv.DictReader(f)
        if "time_s" not in reader.fieldnames or "power_w" not in reader.fieldnames:
            raise ValueError("CSV must have columns: time_s,power_w")

        for row in reader:
            # Handle possible spaces
            t = float(row["time_s"].strip())
            p = float(row["power_w"].strip())
            samples.append(PowerSample(time_s=t, power_w=p))

    samples.sort(key=lambda s: s.time_s)
    if len(samples) < 2:
        raise ValueError("Need at least 2 samples to integrate energy.")
    return samples


def integrate_energy_kwh(samples: List[PowerSample]) -> float:
    """
    Integrate power over time using a left Riemann sum.

    Energy (Joules) = sum( P_i * Δt_i )
    kWh = Joules / (1000 * 3600)
    """
    total_joules = 0.0

    for i in range(len(samples) - 1):
        p = samples[i].power_w
        dt = samples[i + 1].time_s - samples[i].time_s
        if dt < 0:
            raise ValueError("Timestamps must be non-decreasing.")
        total_joules += p * dt

    kwh = total_joules / (1000.0 * 3600.0)
    return kwh


def build_power_twin(
    chip_id: str,
    workload_id: str,
    samples: List[PowerSample],
    trace_file: str,
    extra_metadata: Optional[Dict[str, Any]] = None,
) -> Dict[str, Any]:
    duration_s = samples[-1].time_s - samples[0].time_s
    if duration_s <= 0:
        duration_s = 0.0

    avg_power_w = (
        sum(s.power_w for s in samples) / len(samples) if samples else 0.0
    )
    peak_power_w = max(s.power_w for s in samples) if samples else 0.0

    energy_kwh = integrate_energy_kwh(samples)
    per_second_kwh = energy_kwh / duration_s if duration_s > 0 else 0.0

    solar = energy_kwh / SOLAR_KWH
    rays = solar * RAYS_PER_SOLAR

    twin = {
        "version": "tcs-power-twin-v1",
        "chip_id": chip_id,
        "workload_id": workload_id,
        "timestamp_utc": datetime.now(timezone.utc).isoformat(),
        "units": {
            "time": "seconds",
            "power": "watts",
            "energy": "kWh",
            "solar": "Solar",
            "rays": "Solar Rays",
        },
        "profile": {
            "duration_s": duration_s,
            "samples": len(samples),
            "avg_power_w": avg_power_w,
            "peak_power_w": peak_power_w,
        },
        "energy": {
            "total_kwh": energy_kwh,
            "per_second_kwh": per_second_kwh,
        },
        "solar_cost": {
            "solar": solar,
            "rays": rays,
        },
        "source": {
            "trace_type": "time_power_csv",
            "trace_file": os.path.basename(trace_file),
            "integration_method": "left_riemann",
            "assumptions": [
                "Power value held constant until next timestamp",
                "Timestamps are monotonically increasing",
            ],
        },
        "metadata": extra_metadata or {},
    }

    return twin


def parse_metadata_args(meta_args: Optional[List[str]]) -> Dict[str, Any]:
    """
    Parse metadata from --meta key=value pairs.
    Example:
      --meta process_node_nm=5 voltage_v=0.85 clock_ghz=2.4
    """
    metadata: Dict[str, Any] = {}
    if not meta_args:
        return metadata

    for pair in meta_args:
        if "=" not in pair:
            continue
        key, value = pair.split("=", 1)
        key = key.strip()
        value = value.strip()

        # Try to coerce into int/float where reasonable
        if value.replace(".", "", 1).isdigit():
            # Contains only digits and at most one dot
            if "." in value:
                try:
                    metadata[key] = float(value)
                    continue
                except ValueError:
                    pass
            else:
                try:
                    metadata[key] = int(value)
                    continue
                except ValueError:
                    pass

        # Fallback: keep as string
        metadata[key] = value

    return metadata


def main():
    parser = argparse.ArgumentParser(
        description="TC-S Power Twin builder: convert a time/power CSV trace into Solar cost."
    )
    parser.add_argument(
        "csv_path",
        type=str,
        help="Path to CSV file with columns: time_s,power_w",
    )
    parser.add_argument(
        "--chip-id",
        type=str,
        default="unknown-chip",
        help="Logical chip identifier (e.g., open-eda-cpu-v1).",
    )
    parser.add_argument(
        "--workload-id",
        type=str,
        default="unknown-workload",
        help="Workload identifier (e.g., resnet50-inference-01).",
    )
    parser.add_argument(
        "--out",
        type=str,
        default="",
        help="Optional output JSON file path. If omitted, prints to stdout.",
    )
    parser.add_argument(
        "--meta",
        nargs="*",
        help="Optional metadata key=value pairs. Example: --meta process_node_nm=5 voltage_v=0.85",
    )

    args = parser.parse_args()

    samples = load_power_trace_csv(args.csv_path)
    metadata = parse_metadata_args(args.meta)
    twin = build_power_twin(
        chip_id=args.chip_id,
        workload_id=args.workload_id,
        samples=samples,
        trace_file=args.csv_path,
        extra_metadata=metadata,
    )

    twin_json = json.dumps(twin, indent=2)

    if args.out:
        with open(args.out, "w") as f:
            f.write(twin_json)
        print(f"Power Twin written to {args.out}")
    else:
        print(twin_json)


if __name__ == "__main__":
    main()

File: server/routes/power-twin.ts
import { Router } from 'express';
import multer from 'multer';

const router = Router();

// TC-S Constants
const SOLAR_KWH = 4913.0;
const RAYS_PER_SOLAR = 10000.0;

interface PowerSample {
  time_s: number;
  power_w: number;
}

interface PowerTwinResult {
  version: string;
  chip_id: string;
  workload_id: string;
  timestamp_utc: string;
  units: {
    time: string;
    power: string;
    energy: string;
    solar: string;
    rays: string;
  };
  profile: {
    duration_s: number;
    samples: number;
    avg_power_w: number;
    peak_power_w: number;
  };
  energy: {
    total_kwh: number;
    per_second_kwh: number;
  };
  solar_cost: {
    solar: number;
    rays: number;
  };
  source: {
    trace_type: string;
    trace_file: string;
    integration_method: string;
    assumptions: string[];
  };
  metadata: Record<string, any>;
}

function parseCSV(content: string): PowerSample[] {
  const lines = content.trim().split('\n');
  if (lines.length < 2) {
    throw new Error('CSV must have header and at least one data row');
  }

  const header = lines[0].toLowerCase().split(',').map(h => h.trim());
  const timeIndex = header.findIndex(h => h === 'time_s');
  const powerIndex = header.findIndex(h => h === 'power_w');

  if (timeIndex === -1 || powerIndex === -1) {
    throw new Error('CSV must have columns: time_s, power_w');
  }

  const samples: PowerSample[] = [];
  for (let i = 1; i < lines.length; i++) {
    const line = lines[i].trim();
    if (!line) continue;
    
    const values = line.split(',').map(v => v.trim());
    const time_s = parseFloat(values[timeIndex]);
    const power_w = parseFloat(values[powerIndex]);

    if (isNaN(time_s) || isNaN(power_w)) {
      throw new Error(`Invalid data on line ${i + 1}: ${line}`);
    }

    samples.push({ time_s, power_w });
  }

  samples.sort((a, b) => a.time_s - b.time_s);

  if (samples.length < 2) {
    throw new Error('Need at least 2 samples to integrate energy');
  }

  return samples;
}

function integrateEnergyKwh(samples: PowerSample[]): number {
  let totalJoules = 0;

  for (let i = 0; i < samples.length - 1; i++) {
    const p = samples[i].power_w;
    const dt = samples[i + 1].time_s - samples[i].time_s;
    
    if (dt < 0) {
      throw new Error('Timestamps must be monotonically increasing');
    }
    
    totalJoules += p * dt;
  }

  return totalJoules / (1000.0 * 3600.0);
}

function buildPowerTwin(
  chipId: string,
  workloadId: string,
  samples: PowerSample[],
  traceFile: string,
  metadata: Record<string, any> = {}
): PowerTwinResult {
  const duration_s = samples[samples.length - 1].time_s - samples[0].time_s;
  const avg_power_w = samples.reduce((sum, s) => sum + s.power_w, 0) / samples.length;
  const peak_power_w = Math.max(...samples.map(s => s.power_w));
  const energy_kwh = integrateEnergyKwh(samples);
  const per_second_kwh = duration_s > 0 ? energy_kwh / duration_s : 0;
  const solar = energy_kwh / SOLAR_KWH;
  const rays = solar * RAYS_PER_SOLAR;

  return {
    version: 'tcs-power-twin-v1',
    chip_id: chipId,
    workload_id: workloadId,
    timestamp_utc: new Date().toISOString(),
    units: {
      time: 'seconds',
      power: 'watts',
      energy: 'kWh',
      solar: 'Solar',
      rays: 'Solar Rays'
    },
    profile: {
      duration_s,
      samples: samples.length,
      avg_power_w,
      peak_power_w
    },
    energy: {
      total_kwh: energy_kwh,
      per_second_kwh
    },
    solar_cost: {
      solar,
      rays
    },
    source: {
      trace_type: 'time_power_csv',
      trace_file: traceFile,
      integration_method: 'left_riemann',
      assumptions: [
        'Power value held constant until next timestamp',
        'Timestamps are monotonically increasing'
      ]
    },
    metadata
  };
}

const upload = multer({ 
  storage: multer.memoryStorage(),
  limits: { fileSize: 10 * 1024 * 1024 }
});

router.post('/analyze', upload.single('trace'), async (req, res) => {
  try {
    const chipId = req.body.chip_id || 'unknown-chip';
    const workloadId = req.body.workload_id || 'unknown-workload';
    const metadata: Record<string, any> = {};

    if (req.body.process_node_nm) metadata.process_node_nm = parseFloat(req.body.process_node_nm);
    if (req.body.voltage_v) metadata.voltage_v = parseFloat(req.body.voltage_v);
    if (req.body.clock_ghz) metadata.clock_ghz = parseFloat(req.body.clock_ghz);
    if (req.body.notes) metadata.notes = req.body.notes;

    if (!req.file) {
      return res.status(400).json({ 
        success: false, 
        error: 'No trace file uploaded. Please upload a CSV file with columns: time_s, power_w' 
      });
    }

    const csvContent = req.file.buffer.toString('utf-8');
    const samples = parseCSV(csvContent);
    const powerTwin = buildPowerTwin(chipId, workloadId, samples, req.file.originalname, metadata);

    res.json({
      success: true,
      power_twin: powerTwin
    });
  } catch (error: any) {
    console.error('[Power Twin] Error:', error);
    res.status(400).json({
      success: false,
      error: error.message || 'Failed to process power trace'
    });
  }
});

router.post('/calculate', async (req, res) => {
  try {
    const { samples, chip_id, workload_id, metadata } = req.body;

    if (!samples || !Array.isArray(samples) || samples.length < 2) {
      return res.status(400).json({
        success: false,
        error: 'Must provide at least 2 power samples as array of {time_s, power_w}'
      });
    }

    const validSamples: PowerSample[] = samples.map((s: any) => ({
      time_s: parseFloat(s.time_s),
      power_w: parseFloat(s.power_w)
    }));

    const powerTwin = buildPowerTwin(
      chip_id || 'unknown-chip',
      workload_id || 'unknown-workload',
      validSamples,
      'api-input',
      metadata || {}
    );

    res.json({
      success: true,
      power_twin: powerTwin
    });
  } catch (error: any) {
    console.error('[Power Twin] Error:', error);
    res.status(400).json({
      success: false,
      error: error.message || 'Failed to calculate power twin'
    });
  }
});

router.get('/constants', (req, res) => {
  res.json({
    solar_kwh: SOLAR_KWH,
    rays_per_solar: RAYS_PER_SOLAR,
    description: '1 Solar = 4,913 kWh of renewable energy. 1 Solar = 10,000 Solar Rays.',
    version: 'tcs-power-twin-v1'
  });
});

// External Simulator Integration
const OSS_SIMULATOR_URL = 'https://open-source-eda-tdfranklin101.replit.app';

router.get('/simulator/status', async (req, res) => {
  const simulatorInfo = {
    name: 'Open Silicon Stack',
    url: OSS_SIMULATOR_URL,
    features: [
      'VexRiscv RISC-V Core Simulation',
      'OpenRAM Memory Generator',
      'Skywater 130nm PDK',
      'OpenLane RTL-to-GDSII Flow'
    ],
    power_trace_format: 'CSV with time_s, power_w columns',
    integration: 'embedded_iframe'
  };

  try {
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), 5000);

    const response = await fetch(OSS_SIMULATOR_URL, {
      method: 'HEAD',
      signal: controller.signal
    });
    clearTimeout(timeoutId);

    if (response.ok) {
      res.json({
        success: true,
        simulator: {
          ...simulatorInfo,
          status: 'online'
        }
      });
    } else {
      res.status(503).json({
        success: false,
        error: `Simulator returned status ${response.status}`,
        simulator: {
          ...simulatorInfo,
          status: 'degraded'
        }
      });
    }
  } catch (error: any) {
    const errorMessage = error.name === 'AbortError' ? 'Connection timeout' : 'Connection failed';
    res.status(503).json({
      success: false,
      error: errorMessage,
      simulator: {
        ...simulatorInfo,
        status: 'offline'
      }
    });
  }
});

router.get('/simulator/info', (req, res) => {
  res.json({
    simulator: {
      name: 'Open Silicon Stack',
      description: 'Open-source EDA digital twin chip simulator showcasing VexRiscv, OpenRAM, Skywater PDK, and OpenLane',
      url: OSS_SIMULATOR_URL,
      supported_architectures: [
        { name: 'VexRiscv', type: 'RISC-V Core', description: 'Configurable 32-bit RISC-V processor core' },
        { name: 'OpenRAM', type: 'Memory', description: 'Open-source static RAM compiler' },
        { name: 'Skywater 130nm', type: 'PDK', description: 'Open-source process design kit' },
        { name: 'OpenLane', type: 'Flow', description: 'Automated RTL-to-GDSII synthesis flow' }
      ],
      workflow: [
        'Select chip architecture or configure custom design',
        'Choose workload benchmark',
        'Run power simulation',
        'Export power trace as CSV',
        'Upload to Power Twin for Solar cost calculation'
      ],
      csv_format: {
        required_columns: ['time_s', 'power_w'],
        time_unit: 'seconds',
        power_unit: 'watts',
        example: 'time_s,power_w\\n0.0,0.5\\n0.001,0.6\\n0.002,0.55'
      }
    },
    power_twin: {
      version: 'tcs-power-twin-v1',
      solar_conversion: '1 Solar = 4,913 kWh',
      rays_conversion: '1 Solar = 10,000 Rays',
      integration_method: 'left_riemann'
    }
  });
});

export default router;
